# Import libraries

In [110]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing data

In [111]:
path = r"C:\Users\fardi\OneDrive\Desktop\CF\Data Immersion\Python Fundamentals for Data Analysts\06-09-2021 Instacart Basket Analysis"

In [112]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [113]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [114]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [115]:
df_ords = df_ords.drop('Unnamed: 0',1)

In [116]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [117]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [118]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Mixed datatype

In [119]:
# create a dataframe
df_test = pd.DataFrame()

In [120]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [121]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [122]:
# check for mixed type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [123]:
df_test[[col]].iloc[0]

mix    a
Name: 0, dtype: object

In [124]:
df_test.columns.tolist()

['mix']

In [125]:
# convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

### Finding Missing Values

In [126]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [127]:
# create a subset of the dataframe containing only missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [128]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values


##### There are a few ways to deal with missing data:

- Create a new variable that acts like a flag based on the missing value.
- Impute the value with the mean or median of the column (if the variable is numeric).
- Remove or filter out the missing data.

In [129]:
# use the df.describe() function to find the mean of the column in question
# df['column with missings'].fillna(mean value, inplace=True)

In [130]:
# Use the df_prods.median() function to find median
# df['column with missings'].fillna(median value, inplace=True)

In [131]:
# number of rows in current dataframe 
df_prods.shape

(49693, 5)

In [132]:
# subset of non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [133]:
df_prods_clean.shape

(49677, 5)

In [134]:
# Dropping missing values -inplace(True)->overite dataframe, 
# inplace(False)-> will only return a view of the changed dataframe, leaving the original dataframe untouched
# df_prods[‘product_name’].dropna(inplace = True)

# Duplicates

In [135]:
# look for full duplicates within dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [136]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### Addressing Duplicates

In [137]:
# Number of rows in dataFrame
df_prods_clean.shape

(49677, 5)

In [138]:
# Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [139]:
# Number of rows in dataFrame after dropping duplicates

df_prods_clean_no_dups.shape

(49672, 5)

# Tidying Up and Exporting Changes

In [140]:
# Export dataFrame
# df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

### Question 2 - Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further

In [141]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

###### The Maximum price is a bit odd, mean is 9.9, 50% is 7.1 and 75% is 11.2 which seem to be fine and consistent, but maximun is 99999 which seem too hight and inconsistent.

### Question 3 - Check for mixed-type data in your df_ords dataframe.

In [142]:
# check for mixed type
for col in df_ords.columns.tolist():
  weirdo = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weirdo]) > 0:
    print (col)

##### There are no mixed type in the df_ords DataFrame, thats why nothing is printed

###  Question 5 - Run a check for missing values in your df_ords dataframe.

In [143]:
# finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### days_since_prior_order has 206209 missing values, this may be due that they are new customers, who are ordering for the first time


### Question 6 - Address the missing values using an appropriate method.

#### I don't think deleting the data would be a good idea, as these would delete all information regarding new customer. We could impute them with 0


In [144]:
# Imputing missing values with 0
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [147]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### Question 7 - Run a check for duplicate values in your df_ords data.

In [149]:
# look for full duplicates within dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [150]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### There are no duplicates in the dataFrame

### Quesiton 9 - Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [152]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)

In [48]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [ ]:
# need to fix extreme values
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan